core/src/main/scala/io/projectglow/gff/GffDataSource.scala

```diff
@@ -0,0 +1,245 @@
+    path match {
+      case Some(p) =>
+        val requiredSchema = if (schema == null) {
+          inferSchema (sqlContext, p)
+        } else {
+          schema
+        }
+
+        new GffResourceRelation(sqlContext, p, requiredSchema)
+      case _ => throw new IllegalArgumentException("Path is required")
+    }
+  }
+
+}
+
+
+class GffResourceRelation(
+  val sqlContext: SQLContext,
+  path: String,
+  requiredSchema: StructType
+) extends BaseRelation with TableScan  { // TODO: investigate use of PrunedFilteredScan
+
+  private val spark = sqlContext.sparkSession
+
+
+  override def schema: StructType = {
+    requiredSchema
+  }
+
+
+  override def buildScan(): RDD[Row] = {
+    val csvDf = spark
+      .read
+      .option("sep", COLUMN_DELIMITER)
+      .schema(GffDataSource.gffCsvSchema)
+      .csv(path)
+
+    val dfWithMappedAttributes = convertAttributesToMap(
+      filterCommentsAndFasta(
+        csvDf
+      )
+    )
+
+    requiredSchema.foldLeft(dfWithMappedAttributes)((df, f) =>
+      f match {
+        case f if structFieldsEqualExceptNullability(f, startField) =>
+          df.withColumn(
+            f.name,
+            col(f.name).cast(f.dataType) - 1
+          )
+        case f if structFieldsEqualExceptNullability(f, endField) =>
+          df.withColumn(
+            f.name,
+            col(f.name).cast(f.dataType)
+          )
+        case f if structFieldsEqualExceptNullability(f, scoreField) |
+          structFieldsEqualExceptNullability(f, strandField) |
+          structFieldsEqualExceptNullability(f, phaseField) =>
+          df.withColumn(
+            f.name,
+            when(
+              col(f.name) =!= lit(NULL_IDENTIFIER),
+              col(f.name).cast(f.dataType)
+            ).otherwise(lit(null))
+          )
+        case f => // TODO: To be completed (also rectify column names)
+          df.withColumn(
+            f.name,
+            element_at(col(attributesMapColumnName), f.name)
+          )
+      }
+    )
+      .drop(attributesField.name, attributesMapColumnName) // TODO: Complete the option of keeping attributes field
+      .rdd
+  }
+}
+
+object GffDataSource {
+
+  def inferSchema(sqlContext: SQLContext, path: String): StructType = {
+    val spark = sqlContext.sparkSession
+    val csvDf = spark
+      .read
+      .option("sep", COLUMN_DELIMITER)
+      .schema(GffDataSource.gffCsvSchema)
+      .csv(path)
+
+
+    val attributeTags = convertAttributesToMap(
+      filterCommentsAndFasta(
+        csvDf
+      )
+    )
+      .select(attributesMapColumnName)
+      .withColumn(
+        attributesMapColumnName,
+        explode(
+          map_keys(
+            col(attributesMapColumnName)
+          )
+        )
+      )
+      .agg(
+        collect_set(
+          attributesMapColumnName
+        )
+      )
+      .collect()(0).getAs[Seq[String]](0)
+
+    // generate the schema. The field names will be case and underscore insensitive
```

Comment:
Code style thing: I think this would be more concise and clear if you normalized all the names together and then used [sortBy](https://www.scala-lang.org/api/2.12.0/scala/collection/Seq.html#sortBy[B](f:A=%3EB)(implicitord:scala.math.Ordering[B]):Repr) to put them in the order you want.

Reply:
Revised.